### DISCUSSION TITLE
Send file as response

### DISCUSSION THREAD

**Discussion by Corouna** [vercel/next.js#15453] [Help] [answered]
As titled, is this possible?

Using nodejs + express, i can do it like this
```
res.sendFile(absolute_path_to_file)
```

But i see that Next's response doesn't have sendFile method, and i tried the above like the following

```
res.send(absolute_path_to_file)
```

But it didn't work.

My idea is to serve an image file from a folder located on same level as public folder in NextJS directory and return it as response to the user.  But im not sure if res.send() can do that.

**Comment by jamesmosier** [comment, marked as answer]
Hi @Corouna. The `res` object is Node's [built in response object](https://nodejs.org/api/http.html#http_class_http_serverresponse), so if you wanted to send a file you could follow [something like this example](https://stackoverflow.com/a/10047018/1554860)

```js
    var filePath = path.join(__dirname, 'myfile.mp3');
    var stat = fileSystem.statSync(filePath);

    response.writeHead(200, {
        'Content-Type': 'audio/mpeg',
        'Content-Length': stat.size
    });

    var readStream = fileSystem.createReadStream(filePath);
    // We replaced all the event handlers with a simple call to readStream.pipe()
    readStream.pipe(response);
```

Also be sure to access the file via `process.cwd()` as described here https://nextjs.org/docs/basic-features/data-fetching#reading-files-use-processcwd

**Reply by kachkaev** [reply]
Instead of `readStream.pipe(response);`, it’s better to have

```ts
 await new Promise(function (resolve) {
    downloadStream.pipe(res)
    downloadStream.on('end', resolve)
  })
  ```
  
  Otherwise, the following warning will show up:
  
```txt
  API resolved without sending a response for /api/some/route, this may result in stalled requests.
```

Note that the whole API method needs to be async for that to work:

```ts
export default async (req, res) => {...}
//             ↑↑↑↑
```

**Reply by gajendhir** [reply]
I am sorry but this does not work.

I am have `nextjs 13.5.3` + `reactjs 18.2.0`

The errors received are...

- TypeError: res.status is not a function
- TypeError: res.setHeader is not a function
- TypeError: res.send is not a function

The code that I have use is 

```js
import { readFileSync } from "fs";

export function GET(req: Request, res: Response) {
    const defPath = `e:\\anshlab\\upload`
    const q = new URL( req.url )

    const path = q.searchParams.get('path')
    const img = q.searchParams.get('img')
    const file = `${defPath}\\${path}\\${img}`
    const imgBuf = readFileSync(file)
    console.log(file);
    try {
        const imageBuffer = readFileSync(file);
    
        res.setHeader('Content-Type', 'image/png');
        res.send(imageBuffer);
      } catch (e: any) {
        console.log(e.message);
        res.status(400).json({ error: true, message: 'Image not found' });
      }
    //console.log(res);
}
```

**Reply by eric-burel** [reply]
@gajendhir you are conflating:
- Next.js API routes, available since Next version 9, defined in `pages/api` folder, and that use Next.js API (similar to Express)
- Next.js Route Handlers available since Next 13, defined in `app` folder with files named `route`, and that use the "Web Platform" API (same types as `fetch`)

Your code is using API routes syntax in a route handler. You may either want to use [the route handler syntax from answer below](https://github.com/vercel/next.js/discussions/15453#discussioncomment-6748645) or use an API route.

**Comment by cristiand391** [comment]
It's technically possible, but I'm not sure if it will work on all deployment platforms.

As @jamesmosier said, you can send a buffer using `res.send(buffer)`, but keep in mind that you will be reading files from outside Next's default folders, so I wouldn't be surprised if it didn't work outside your dev environment.

Example:
```javascript
/*
Project structure:
.
├── images_folder
│   └── next.jpg
├── package.json
├── pages
│   ├── api
│   │   └── image.js
│   └── index.js
├── README.md
└── yarn.lock
*/

// pages/api/image.js

import fs from 'fs'
import path from 'path'

const filePath = path.resolve('.', 'images_folder/next.jpg')
const imageBuffer = fs.readFileSync(filePath)

export default function(req, res) {
  res.setHeader('Content-Type', 'image/jpg')
  res.send(imageBuffer)
}
```

Out of curiosity, why do you need to serve files outside of Next's `public` folder?

**Reply by jamesmosier** [reply]
`process.cwd()` would work to serve local static files outside of Next's directories https://nextjs.org/docs/basic-features/data-fetching#reading-files-use-processcwd

**Reply by cristiand391** [reply]
Already tested and it works :100: 

It seems using `process.cwd()` is the only way to do it without breaking things, see #8251.

**Reply by Corouna** [reply]
> Out of curiosity, why do you need to serve files outside of Next's `public` folder?

Coz the image is dynamically generated... user send the b64 file and the API will process and write the file inside the destination folder. Then i will return back the processed file back to the user, and delete the file from the destination folder.

**Comment by Corouna** [comment]
@jamesmosier @cristiand391 thanks for the reply. Im still on vacation now but will definitely give that a try once im back. Many thanks again 😄

**Reply by mfbx9da4** [reply]
My use case is for safari push notifications. I have to serve a zip file from a POST request. Static files won't cut it as it's not possible to serve `POST`. I will try serving the file from an api route instead

https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/NotificationProgrammingGuideForWebsites/PushNotifications/PushNotifications.html @Corouna

**Comment by surjithctly** [comment]
For future googlers, Here's the full working code. 

Live demo: https://stackblitz.com/edit/nextjs-image-api?file=pages/api/[id].js
Try adding  /api/1 to /api/6 in the URL

```js
import fs from 'fs';
import path from 'path';

export default function(req, res) {
  const id = req.query.id;
  const filePath = path.join(process.cwd(), `/public/images/${id}.svg`);
  try {
    const imageBuffer = fs.readFileSync(filePath);
    res.setHeader('Content-Type', 'image/svg+xml');
    res.send(imageBuffer);
  } catch (e) {
    res.status(400).json({ error: true, message: 'Image not found' });
  }
}

```

**Reply by timneutkens** [reply]
In general it's recommended to pipe the stream instead of reading a file synchronously (this will block every request from processing as it's not async). As shown in: https://github.com/vercel/next.js/discussions/15453#discussioncomment-220281

**Reply by surjithctly** [reply]
@timneutkens I might be doing something wrong, but it doesn't work. It just returns empty.

```js
import fs from 'fs';
import path from 'path';

export default async function(req, res) {
  const id = req.query.id;
  const filePath = path.join(process.cwd(), `/public/images/${id}.svg`);
  try {
    const imageBuffer = fs.createReadStream(filePath);
    await new Promise(function(resolve) {
      res.setHeader('Content-Type', 'image/svg+xml');
      imageBuffer.pipe(res);
      imageBuffer.on('end', resolve);
    });
  } catch (e) {
    res.status(400).json({ error: true, message: e });
  }
}
```

Can you tell me what's wrong here? I'm new to this.

**Reply by surjithctly** [reply]
It works now :)

Added error handling for `ENOENT: no such file or directory, open` as well. 

```js
import fs from 'fs';
import path from 'path';

export default async function(req, res) {
  const id = req.query.id;

  try {
    const filePath = path.join(process.cwd(), `/public/images/${id}.svg`);
    const imageBuffer = fs.createReadStream(filePath);

    await new Promise(function(resolve) {
      res.setHeader('Content-Type', 'image/svg+xml');
      imageBuffer.pipe(res);
      imageBuffer.on('end', resolve);
      imageBuffer.on('error', function(err) {
        if (err.code === 'ENOENT') {
          res.status(400).json({
            error: true,
            message: 'Sorry we could not find the file you requested!'
          });
          res.end();
        } else {
          res
            .status(500)
            .json({ error: true, message: 'Sorry, something went wrong!' });
          res.end();
        }
      });
    });
  } catch (err) {
    res.status(400).json({ error: true, message: err });
    res.end();
  }
}
```

**Comment by 1finedev** [comment]
Could someone please show a working example of how to save this data on the client side pages. i keep getting a corrupted file with this method 

```js
    const data = await axios.get("/api/shipments/downloadManifest");
    const pdfBlob = new Blob([data.data], { type: "application/xlsx" });
    return saveAs(pdfBlob, "manifest.xlsx");
```

**Reply by 1finedev** [reply]
adding
```js 
{ responseType: "blob" }
```
fixed it for me 

```js
const data = await axios.get("/api/shipments/downloadManifest", {
      responseType: "blob",
    });
const pdfBlob = new Blob([data.data], { type: "application/xlsx" });
return saveAs(pdfBlob, "manifest.xlsx");
```

**Comment by m10rten** [comment]
Hi all, been trying to get an API route to send an image based on a URL within a database object.

code is like this:
```js
export default async function handler(req, res) {
  const { id } = req.query;
  let canSeeImage = false; // there is more behind this, but its not relevant.
  const image = await get(id);

  if(canSeeImage) {

    const imageRes = fetch(image.url);
  
    // how to send it back?
  
    // tried this:
    // const blob = await imageResponse.blob();
    // const blobUrl = URL.createObjectURL(blob);
    // const imageBlob = new Blob([blob], { type: 'image/jpeg' });
    // res.setHeader('Content-Type', imageBlob.type);
    // res.setHeader('Content-Length', imageBlob.size);
    // res.setHeader('Cache-Control', 'public, max-age=31536000, immutable');
    // res.status(200).send(imageBlob);
  }
}
```

**Reply by m10rten** [reply]
Alright, as always: stackoverflow comes in clutch:

Found this for a simple node express app, and it works for Next as well.

Requires the writehead 200 otherwise next will think it has not send a response.
```js
res.writeHead(200, { 'Content-Type': 'image/jpeg' });

if (image.url.startsWith('https'))
  https.get(image.url, (stream) => {
    stream.pipe(res);
  });
else {
  http.get(image.url, (stream) => {
    stream.pipe(res);
  });
}
```

**Reply by vasyl-zubach** [reply]
@m10rten in case you still want to still use `fetch` API for that, you can use:

```javascript
 const imageResponse = await fetch(image.url);

  if (!imageResponse.ok) {
    return res.status(404);
  }

  res.writeHead(200, {
    'Content-Type': `image/jpeg`,
    'Cache-Control': `public, immutable, no-transform, s-maxage=31536000, max-age=31536000`
  });

  res.end(Buffer.from(await imageResponse.arrayBuffer()));
```

**Reply by m10rten** [reply]
Thanks for your answer, however I'm satisfied with my solution and it works perfectly. 
No need to re do it over for the sake of redoing it.

**Comment by AaronMeyers** [comment]
Does anyone know a solution that works with the newer App router? API routes are different and don’t give you an res object and as far as i can tell, require you to return a NextReponse object. I need to pipe a video through an api route and I can’t figure out how to do it with this App router API routes…

**Reply by jianyuan** [reply]
@AaronMeyers You can set the body in the constructor:

```ts
// before
res.status(200).setHeader('content-type', 'image/png').send(data);

// after
return new Response(data, { headers: { 'content-type': 'image/png' } });
// or
const response = new NextResponse(data)
response.headers.set('content-type', 'image/png');
return response;
```

**Reply by Lucifergene** [reply]
Finally, after hours of exploring Stack Overflow, got the solution :)

**Reply by eric-burel** [reply]
Hi, great response for usual scenarios however it's not great for big files: you have to load the whole file in memory (the `data` variable) and then send it. Instead what you want is to create a read stream and pipe it in the reponse: the file is sent as it is read, avoiding to bloat your server RAM.

I'll try to craft a solution, if that's possible.

`Response` constructor (so by extension `NextResponse`) accepts a `ReadableStream` object as its body.

Here is a [Stack overflow post about how to go from fs.ReadStream to ReadableStream](https://stackoverflow.com/questions/70905931/how-to-convert-readablestream-into-readstream-in-nodejs). What we need here is the opposite.

I've [just opened a new question on Stack Overflow](https://stackoverflow.com/questions/76763053/how-to-convert-readstream-into-readablestream-in-nodejs) for the conversion we want here.

**Reply by rohithandique** [reply]
@eric-burel hi Eric, just saw your post on stack overflow. I'm trying to use `readFile` in an api route and then send the data back. Facing the same weird 'ECONNRESET' issue you're facing (same stack trace but the message is `socket hang up`).
Also you mentioned here, NextResponse accepts a ReadableStream object as its body but how do I pipe it in the response? I'm not able to figure out the syntax for it. Right now I'm loading the data in a variable (which isnt a good thing as you mentioned above).

**Reply by eric-burel** [reply]
Thanks for the insights, so this is not just me. We need to cut Next.js code open to see how it handle the `Response` object returned by the route handler, perhaps it doesn't support streams. The fact that `NextResponse` accepts a stream is because it inherits `Response` which is an object initially designed for the browser (so you can read the response as a stream), not really the server.

**Reply by eric-burel** [reply]
See relevant Reddit thread but with no clear answer: https://www.reddit.com/r/nextjs/comments/13toeob/nextjs_response_with_a_stream/

I found some docs though with a demo using OpenAI: https://nextjs.org/docs/app/building-your-application/routing/router-handlers#streaming
But not fs. I've looked for "how to read a file using an iterator in Node" but didn't yet find a working piece of code.

**Reply by rohithandique** [reply]
@eric-burel the socket error you mentioned on SO might also be because of this - https://nextjs.org/docs/messages/api-routes-response-size-limit. API Routes apparently have a 4mb limit, and anything above that throws out an error. I've tried sending files using `readFile` that are above that limit, and after sending 1 file, it throws that error and crashes the application. But for multiple files, smaller than that, it is working fine.
Again, it is mentioned that the issue only exists if you use the API route in a edge runtime environment but Next uses node runtime as default environment, which I'm using, but it still throws the error for large files. Even configuring the limit and increasing it is an option but using it in Nextjs 13 with App Router is throwing me a deprecation warning so I can't increase the limit either.
Also, the socket hang up issue might be caused because your app crashed due to high memory usage (this doesnt happen locally usually, at least for me, but on a production environment inside a docker container). Links to issues - https://github.com/vercel/next.js/issues/49929 and https://github.com/vercel/next.js/issues/49677

**Reply by eric-burel** [reply]
Interesting, in my case I was sending a 344b file so definitely under the threshold but perhaps my stream never correctly ends since the syntax is messed up. I wonder if this limitation applies to streams? Since the whole point is to not send 4mb in one go. But perhaps this is summed over the file size. Anyway I am below this threshold here + in a Node.js environment so it's perhaps a different issue.

**Reply by rohithandique** [reply]
@eric-burel were you able to figure out how to pipe a stream to NextResponse object though?
creating a new NextResponse object and piping into it throws an error for me since it's not a pipeable object :)

I also increased my container memory to 500 mb from 200mb, and coded a limit to the max size of a file which can be transferred (just 4 mb) from an api route, and it has been working fine for now. I'll keep on monitoring and increasing limits. but the best case scenario is to pipe the data to a NextResponse itself :(

**Reply by eric-burel** [reply]
@rohithandique made some progress on the stream part, see https://stackoverflow.com/a/76780381/5513532
This works in a tiny script when printing the response. However in a Next.js Route Handler, using Node.js, I now have this error:
```
- error TypeError: This ReadableStream did not return bytes.
    at consumeUint8ArrayReadableStream (/code/devographics/monorepo/node_modules/.pnpm/next@13.4.8_e7wxmg5kx5hzmlsglkcl4if3fi/node_modules/next/dist/compiled/edge-runtime/index.js:1:1111242)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async sendResponse (/code/devographics/monorepo/node_modules/.pnpm/next@13.4.8_e7wxmg5kx5hzmlsglkcl4if3fi/node_modules/next/dist/server/send-response.js:42:34)
    at async doRender (/code/devographics/monorepo/node_modules/.pnpm/next@13.4.8_e7wxmg5kx5hzmlsglkcl4if3fi/node_modules/next/dist/server/base-server.js:1032:21)
    at async cacheEntry.responseCache.get.incrementalCache.incrementalCache (/code/devographics/monorepo/node_modules/.pnpm/next@13.4.8_e7wxmg5kx5hzmlsglkcl4if3fi/node_modules/next/dist/server/base-server.js:1197:28)
    at async /code/devographics/monorepo/node_modules/.pnpm/next@13.4.8_e7wxmg5kx5hzmlsglkcl4if3fi/node_modules/next/dist/server/response-cache/index.js:99:36

-  ┌ GET /api/data-export?surveyId=css2021&editionId=css2021 500 in 206ms
   │
   └──── POST http://localhost:8080/ 200 in 10ms (cache: MISS)

- error Error: aborted
    at connResetException (node:internal/errors:711:14)
    at Socket.socketCloseListener (node:_http_client:454:19)
    at Socket.emit (node:events:525:35)
    at TCP.<anonymous> (node:net:313:12) {
  code: 'ECONNRESET'
}
```
The ECONNRESET error is probably just a consequence of the first error.

Related: https://github.com/vercel/next.js/issues/38736

(@leerob if by any chance we could have an official answer on this this would be awesome :pray:)

**Comment by eric-burel** [comment]
Below is an updated answer for Next 13 Route Handlers. This answer uses streams, in order to avoid bloating your server RAM for big files or high request volumes.

You may get more information about the conversion from Node.js logic (Next.js API routes) to web platform logic used by route handlers (needed so they work both in Node.js and the Edge runtime) in [this Stack Overflow answer](https://stackoverflow.com/a/76780381/5513532).

See an example usage in the [State of JavaScript codebase](https://github.com/Devographics/Monorepo/blob/e1179268188111380ccaaa0a134318273d9f6594/surveyadmin/src/app/api/data-export/route.ts), to download a CSV+JSON zip out of a Mongo database. In particular, you'll want to set the "Content-Size" header properly with the file size.

Bonus: we have also tried to convert `Request` to `NextAPIRequest`, in order to be able to reuse Express/Connect logic in route handlers, which would enable using Passport namely. It didn't work as this is quite hard to reproduce all the features of Express middlewares [but here is the code of this attempt](https://github.com/Devographics/Monorepo/blob/e1179268188111380ccaaa0a134318273d9f6594/surveyform/src/app/api/nespresso.ts).

Regarding file download:
```ts
// @see https://github.com/vercel/next.js/discussions/15453#discussioncomment-6226391
import fs from "fs"

/**
 * Took this syntax from https://github.com/MattMorgis/async-stream-generator
 * Didn't find proper documentation: how come you can iterate on a Node.js ReadableStream via "of" operator?
 * What's "for await"?
 */
async function* nodeStreamToIterator(stream: fs.ReadStream) {
    for await (const chunk of stream) {
        yield chunk;
    }
}

/**
 * Taken from Next.js doc
 * https://nextjs.org/docs/app/building-your-application/routing/router-handlers#streaming
 * Itself taken from mozilla doc
 * https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#convert_async_iterator_to_stream
 */
function iteratorToStream(iterator): ReadableStream {
    return new ReadableStream({
        async pull(controller) {
            const { value, done } = await iterator.next()

            if (done) {
                controller.close()
            } else {
                // conversion to Uint8Array is important here otherwise the stream is not readable
                // @see https://github.com/vercel/next.js/issues/38736
                controller.enqueue(new Uint8Array(value))
            }
        },
    })
}

export function streamFile(path: string): ReadableStream {
    const downloadStream = fs.createReadStream(path);
    const data: ReadableStream = iteratorToStream(nodeStreamToIterator(downloadStream))
    return data
}
```
Usage:
```ts
        const stats = await fsPromises.stat(zipFilePath);
        const data: ReadableStream = streamFile(zipFilePath)
        const res = new NextResponse(data, {
            status: 200,
            headers: new Headers({
                "content-disposition": `attachment; filename=${path.basename(
                    zipFilePath
                )}`,
                "content-type": "application/zip",
                "content-length": stats.size + "",
            })
        })
        return res
 ```

Thanks @rohithandique for the feedback and the many people whose code I had to copy paste and merge to produce this beauty.

**Comment by Lrios4403** [comment]
Yoo, Wall_Naive here from the reddit question [(https://www.reddit.com/r/nextjs/comments/13toeob/nextjs_response_with_a_stream/](https://www.reddit.com/r/nextjs/comments/13toeob/nextjs_response_with_a_stream/) Heres how I setup the streaming from the file. I just added the events when we got dada, when we ended, and when we got an error. It works with the given .iso file. The given example is with NextJS 13 with Typescript.

Credit to [eric-burel](https://github.com/eric-burel) I couldn't figure out the NextJS Response headers. Still trying to figure out etags

```typescript
import fs, { Stats } from "fs";
import { NextRequest, NextResponse } from "next/server";
import path from "path";
import { ReadableOptions } from "stream";

function streamFile(path: string, options?: ReadableOptions): ReadableStream<Uint8Array> {
    const downloadStream = fs.createReadStream(path, options);

    return new ReadableStream({
        start(controller) {
            downloadStream.on("data", (chunk: Buffer) => controller.enqueue(new Uint8Array(chunk)));
            downloadStream.on("end", () => controller.close());
            downloadStream.on("error", (error: NodeJS.ErrnoException) => controller.error(error));
        },
        cancel() {
            downloadStream.destroy();
        },
    });
}

export async function GET(req: NextRequest): Promise<NextResponse> {
    const uri = req.nextUrl.searchParams.get("uri");
    const file = "/home/manjaro/Downloads/manjaro-kde-22.1.3-230529-linux61.iso"; // req.nextUrl.searchParams.get("file");

    const stats: Stats = await fs.promises.stat(file); 
    const data: ReadableStream<Uint8Array> = streamFile(file);   //Stream the file with a 1kb chunk
    const res = new NextResponse(data, {
        status: 200,
        headers: new Headers({
            "content-disposition": `attachment; filename=${path.basename(file)}`,
            "content-type": "application/iso",
            "content-length": stats.size + "",
        }),
    });

    return res;
}


```

**Reply by rohithandique** [reply]
@Lrios4403 @eric-burel have you seen any improvements regarding the applications memory utilisation in a non-local environment with limited memory when transferring files? I am seeing spikes regardless of what method I'm using, and had to increase memory of my pod to accommodate the application so it doesn't keep crashing.

**Reply by Lrios4403** [reply]
cant say ive compared and contrasted. ive been playing with the `options?: ReadableOptions` and while setting the highWatermark to 1kb makes it slow i havent seen too much ram being eaten up

**Reply by ghost** [reply]
> Yoo, Wall_Naive here from the reddit question [(https://www.reddit.com/r/nextjs/comments/13toeob/nextjs_response_with_a_stream/](https://www.reddit.com/r/nextjs/comments/13toeob/nextjs_response_with_a_stream/) Heres how I setup the streaming from the file. I just added the events when we got dada, when we ended, and when we got an error. It works with the given .iso file. The given example is with NextJS 13 with Typescript.
> 
> Credit to [eric-burel](https://github.com/eric-burel) I couldn't figure out the NextJS Response headers. Still trying to figure out etags
> 
> ```typescript
> import fs, { Stats } from "fs";
> import { NextRequest, NextResponse } from "next/server";
> import path from "path";
> import { ReadableOptions } from "stream";
> 
> function streamFile(path: string, options?: ReadableOptions): ReadableStream<Uint8Array> {
>     const downloadStream = fs.createReadStream(path, options);
> 
>     return new ReadableStream({
>         start(controller) {
>             downloadStream.on("data", (chunk: Buffer) => controller.enqueue(new Uint8Array(chunk)));
>             downloadStream.on("end", () => controller.close());
>             downloadStream.on("error", (error: NodeJS.ErrnoException) => controller.error(error));
>         },
>         cancel() {
>             downloadStream.destroy();
>         },
>     });
> }
> 
> export async function GET(req: NextRequest): Promise<NextResponse> {
>     const uri = req.nextUrl.searchParams.get("uri");
>     const file = "/home/manjaro/Downloads/manjaro-kde-22.1.3-230529-linux61.iso"; // req.nextUrl.searchParams.get("file");
> 
>     const stats: Stats = await fs.promises.stat(file); 
>     const data: ReadableStream<Uint8Array> = streamFile(file);   //Stream the file with a 1kb chunk
>     const res = new NextResponse(data, {
>         status: 200,
>         headers: new Headers({
>             "content-disposition": `attachment; filename=${path.basename(file)}`,
>             "content-type": "application/iso",
>             "content-length": stats.size + "",
>         }),
>     });
> 
>     return res;
> }
> 
> 
> ```
> 

Is it possible to delete / destroy the file after response is sent? Something like `downloadStream.on('end', ()=> {
fs.unlink(filename, ()=>{})
controller.close()
})`

**Reply by eric-burel** [reply]
@shennan-wong-partior you might want to look for discussions talking about doing operations after a response is sent in general, or even open a new issue/discussion. In API routes, or Express for instance, what you can do is send the request but return only later on:
```js
res.send(...) // notice that you don't return here, while in Next.js Route handler you have to return the response so its sent to the end user
fs.unlink(...)
```
But in route handler the thing is that the part that actually send the response is internal to Next.js, you only create a Response object but you don't control when it is sent. No idea how to bypass this yet, you may give a shot at registering an event handler on the ReadableStream end maybe? I'll think about it

**Reply by eric-burel** [reply]
@rohithandique I didn't go as far as testing the actual impact indeed, you might want to open an issue with repro if you do notice a spike even when using a stream. Also make sure to test on a built app (in dev the first call to your API endpoint will provoke a spike but due to the build)

**Reply by XenonyxBlaze** [reply]
@Lrios4403
how do you handle the response ?? (sorry still a newbie to react/next)
what happens with the "res" response object

**Reply by Lrios4403** [reply]
dont worry about asking new questions this isn't StackOverflow. (also im newish so may not be any better)

In Next 13 there are pages.tsx (or jsx or whatever language your using) and route.tsx. With route.tsx you can handle HTTP methods like GET and POST requested from the user and return a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) object. If you want to GET data from a webserver (like a webpage, a file, etc) you can request said file and do whatever you want with it (for example, show a website in html and have the browser render it). In order to POST (or submit data) to the webserver you can make a HTTP request with POST. In any HTTP method function it doesent matter what happens inside said funciton as long as you return a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response)  or a [NextResponse](https://nextjs.org/docs/app/api-reference/functions/next-response). 

TLDR; nextjs allows you to GET and POST to a webserver with `route.tsx` to serve a user a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response)  or a [NextResponse](https://nextjs.org/docs/app/api-reference/functions/next-response).

In the code I provided is code to a `route.tsx` file. For this example I have decided to stream the file _/home/manjaro/Downloads/manjaro-kde-22.1.3-230529-linux61.iso_ for the user to download.

- First, I define my functions; _streamFile_ and _GET_
- Next, I setup _streamFile_ to stream a file of my choice
- Then, I setup _GET_ to return a NextResponse of the stream so the user can download a file
- Finally, wherever your `route.tsx` is in your project (like _/app/api/test/route.tsx_) whenever you access the webpage (like to use _/app/api/test/route.tsx_ you need to go to _localhost:3000/api/test_) Next will return the [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response)  or [NextResponse](https://nextjs.org/docs/app/api-reference/functions/next-response)

```typescript
import fs, { Stats } from "fs";
import { NextRequest, NextResponse } from "next/server";
import path from "path";
import { ReadableOptions } from "stream";

/**
 * Return a stream from the disk
 * @param {string} path - The location of the file 
 * @param {ReadableOptions} options - The streamable options for the stream (ie how big are the chunks, start, end, etc).
 * @returns {ReadableStream} A readable stream of the file
 */
function streamFile(path: string, options?: ReadableOptions): ReadableStream<Uint8Array> {
    const downloadStream = fs.createReadStream(path, options);

    return new ReadableStream({
        start(controller) {
            downloadStream.on("data", (chunk: Buffer) => controller.enqueue(new Uint8Array(chunk)));
            downloadStream.on("end", () => controller.close());
            downloadStream.on("error", (error: NodeJS.ErrnoException) => controller.error(error));
        },
        cancel() {
            downloadStream.destroy();
        },
    });
}

/**
 * Return a NextRequest with the downloaded file
 * @param {NextRequest} req - The HTTP request
 * @param {ReadableOptions} options - The streamable options for the stream (ie how big are the chunks, start, end, etc).
 * @returns {Promise<NextResponse>} A NextResponse with the downloadable file
 */
export async function GET(req: NextRequest): Promise<NextResponse> {
    const uri = req.nextUrl.searchParams.get("uri");                                // Get the `uri` from the search parameters (for example, `https://google.com/download?uri=asdf` will return `asdf`)
    const file = "/home/manjaro/Downloads/manjaro-kde-22.1.3-230529-linux61.iso";   // The manjaro iso file that I want people to download

    const stats: Stats = await fs.promises.stat(file);                              // Get the file size
    const data: ReadableStream<Uint8Array> = streamFile(file);                      // Stream the file with a 1kb chunk
    const res = new NextResponse(data, {                                            // Create a new NextResponse for the file with the given stream from the disk
        status: 200,                                                                    //STATUS 200: HTTP - Ok
        headers: new Headers({                                                          //Headers
            "content-disposition": `attachment; filename=${path.basename(file)}`,           //State that this is a file attachment
            "content-type": "application/iso",                                              //Set the file type to an iso
            "content-length": stats.size + "",                                              //State the file size
        }),
    });

    return res;                                                                     // Return the NextResponse with the file so NextJS can send the file to the user
}
```

**Comment by niklasp** [comment]
I recently ran into the same issue. I want an equivalent to `res.write` where I can still append data even when the stream is returned. That is not possible with app router? Is there no way I can write a `hanlder` function ?

**Comment by gajendhir** [comment]
@eric-burel - Thank you very much for highlighting the difference between the `/pages/api` and the `/app/api` routing strategies. I have both methods up and working. I am now successfully able to deliver image files using the API routes. With this I am now able to deliver files that are stored in database fields, without having to save them to the public folder.

Please help me with the last part of this query... 
- how to use `readfileSync` or `createReadStream`  for a file outside the public folder.

In the current usage, `readfileSync` is return an empty (undefined) `ImageBuffer` and `createReadStream` is also giving similar results.

**Comment by eric-burel** [comment]
Hey folks, I've worked up a simpler solution thanks to @karlhorky feedback, Node.js has a [`readableWebStream`](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#filehandlereadablewebstreamoptions) experimental on filehandles that makes it easier to implement.

Live demo: https://nextpatterns.dev/p/gems/route-handlers-file-streaming
Code sample:
```ts

import fs from "fs/promises"
import path from "path"

export async function GET(request: Request) {
    const filePath = path.resolve("/path-to-your-image.jpg")
    const stats = await fs.stat(filePath);
    const fileHandle = await fs.open(filePath)
    const stream = fileHandle.readableWebStream(
        { type: "bytes" }
    )
    return new Response(stream, {
        status: 200,
        headers: new Headers({
            "content-type": "image/jpeg",
            "content-length": stats.size + "",
        })
    })
    // needs next 15 with next/after
    after(() => { fileHandle.close() })
}
```

In case you have a doubt about perf impact, the first spike on the left is the RAM impact of the proper solution (small spike), the second is the RAM impact of the wrong solution (big spike). Screenshot is actually from an Astro app and without proper file closing but you get the point, in the stream implementation, the big file is not copied in memory, in the copy then stream implementation, it bloats the RAM.

![RAM impact of copying file](https://github.com/user-attachments/assets/477ac9d1-565a-43eb-9f54-1fb46bcba0fc)


See https://github.com/nodejs/node/issues/54041 and https://github.com/vercel/next.js/discussions/68109

**Reply by andrewgreenh** [reply]
You could always do the transformation via the iterator protocol like this:

```ts
import * as fs from "fs";
import { ReadableStream } from "stream/web";

const rs = fs.createReadStream("./package.json");
const ws = ReadableStream.from(rs);

for await (const x of ws) {
  console.log(x.toString());
}
```

Since node readable streams are async iterators, you can use the ReadableStream.from method to get a web stream and this should cover error handling, back pressure and finalization.
